File: claude-code-agents-typescript/src/pipeline/modules.ts
// src/pipeline/modules.ts — Chain-of-thought wrapper modules for each pipeline signature
import type { Result } from '../types.js';
import { Err } from '../types.js';
import type {
  CodegenRouterInputT,
  CodegenRouterOutputT,
  PageClassifierInputT,
  PageClassifierOutputT,
  PluginDesignerInputT,
  PluginDesignerOutputT,
  QualityScorerInputT,
  QualityScorerOutputT,
  SelectorProposerInputT,
  SelectorProposerOutputT,
} from './signatures.js';
import {
  CodegenRouterOutput,
  PageClassifierOutput,
  PluginDesignerOutput,
  QualityScorerOutput,
  SelectorProposerOutput,
} from './signatures.js';

// ── Module Error ────────────────────────────────────────────────
export class ModuleError extends Error {
  constructor(
    readonly module: string,
    message: string,
  ) {
    super(`[${module}] ${message}`);
    this.name = 'ModuleError';
  }
}

// ── Base Module Pattern ─────────────────────────────────────────
type ModuleForward<I, O> = (input: I) => Promise<Result<O, ModuleError>>;

function createModule<I, O>(
  name: string,
  _schema: { parse(data: unknown): O },
  buildPrompt: (input: I) => string,
): ModuleForward<I, O> {
  return async (input: I): Promise<Result<O, ModuleError>> => {
    try {
      const prompt = buildPrompt(input);
      // In production, this calls Anthropic API with chain-of-thought
      // For now, return a structured placeholder that can be overridden
      void prompt;
      return Err(
        new ModuleError(name, 'LM not configured — call pipeline.configure() first'),
      );
    } catch (e) {
      return Err(
        new ModuleError(name, e instanceof Error ? e.message : String(e)),
      );
    }
  };
}

// ── Page Classifier Module ──────────────────────────────────────
export const pageClassifierModule: ModuleForward<
  PageClassifierInputT,
  PageClassifierOutputT
> = createModule('PageClassifier', PageClassifierOutput, (input) =>
  [
    'Classify this web page into one of: doc, research, news, api, legal, product, plugin_spec, sdk_ref',
    `URL: ${input.url}`,
    `Title: ${input.title}`,
    `Content: ${input.contentSnippet.slice(0, 2000)}`,
    `HTML: ${input.htmlSnippet.slice(0, 1000)}`,
  ].join('\n'),
);

// ── Quality Scorer Module ───────────────────────────────────────
export const qualityScorerModule: ModuleForward<
  QualityScorerInputT,
  QualityScorerOutputT
> = createModule('QualityScorer', QualityScorerOutput, (input) =>
  [
    'Score extraction quality (completeness, structure, links) each 0.0-1.0',
    `URL: ${input.url}`,
    `Content: ${input.extractedContent.slice(0, 3000)}`,
    `Structured data: ${input.structuredData.slice(0, 2000)}`,
    `Selectors: ${input.selectorsUsed}`,
    `Links: ${input.linkCount}`,
  ].join('\n'),
);

// ── Selector Proposer Module ────────────────────────────────────
export const selectorProposerModule: ModuleForward<
  SelectorProposerInputT,
  SelectorProposerOutputT
> = createModule('SelectorProposer', SelectorProposerOutput, (input) =>
  [
    'Propose improved selectors. Format: old -> new',
    `Spider: ${input.spiderName}`,
    `Current: ${input.currentSelectors}`,
    `Failing: ${input.failingSelectors}`,
    `HTML: ${input.htmlSample.slice(0, 5000)}`,
    `Page type: ${input.pageType}`,
  ].join('\n'),
);

// ── Plugin Designer Module ──────────────────────────────────────
export const pluginDesignerModule: ModuleForward<
  PluginDesignerInputT,
  PluginDesignerOutputT
> = createModule('PluginDesigner', PluginDesignerOutput, (input) =>
  [
    'Design a Claude Code plugin with skills, agents, connectors',
    `Domain: ${input.domain}`,
    `Summaries: ${input.crawledSummaries}`,
    `Page types: ${input.discoveredPageTypes}`,
    `Existing: ${input.existingPlugins}`,
  ].join('\n'),
);

// ── Codegen Router Module ───────────────────────────────────────
export const codegenRouterModule: ModuleForward<
  CodegenRouterInputT,
  CodegenRouterOutputT
> = createModule('CodegenRouter', CodegenRouterOutput, (input) =>
  [
    'Route codegen to language + scaffold type',
    `Task: ${input.taskDescription}`,
    `Env: ${input.targetEnvironment}`,
    `Preferred: ${input.preferredLanguages}`,
    `Constraints: ${input.constraints}`,
  ].join('\n'),
);

File: claude-code-agents-python/src/dspy_pipeline/modules.py
"""DSPy ChainOfThought wrapper modules for each pipeline signature."""

from __future__ import annotations

import dspy

from .signatures import (
    CodegenRouter,
    PageClassifier,
    PluginDesigner,
    QualityScorer,
    SelectorProposer,
)


class PageClassifierModule(dspy.Module):
    """ChainOfThought wrapper for page classification."""

    def __init__(self) -> None:
        super().__init__()
        self.classify = dspy.ChainOfThought(PageClassifier)

    def forward(
        self,
        url: str,
        title: str,
        content_snippet: str,
        html_snippet: str = "",
    ) -> dspy.Prediction:
        """Classify a page given its URL, title, and content."""
        return self.classify(
            url=url,
            title=title,
            content_snippet=content_snippet[:2000],
            html_snippet=html_snippet[:1000],
        )


class QualityScorerModule(dspy.Module):
    """ChainOfThought wrapper for extraction quality scoring."""

    def __init__(self) -> None:
        super().__init__()
        self.score = dspy.ChainOfThought(QualityScorer)

    def forward(
        self,
        url: str,
        extracted_content: str,
        structured_data: str = "{}",
        selectors_used: str = "",
        link_count: int = 0,
    ) -> dspy.Prediction:
        """Score extraction quality."""
        return self.score(
            url=url,
            extracted_content=extracted_content,
            structured_data=structured_data,
            selectors_used=selectors_used,
            link_count=link_count,
        )


class SelectorProposerModule(dspy.Module):
    """ChainOfThought wrapper for selector improvement proposals."""

    def __init__(self) -> None:
        super().__init__()
        self.propose = dspy.ChainOfThought(SelectorProposer)

    def forward(
        self,
        spider_name: str,
        current_selectors: str,
        failing_selectors: str,
        html_sample: str,
        page_type: str = "doc",
    ) -> dspy.Prediction:
        """Propose improved selectors."""
        return self.propose(
            spider_name=spider_name,
            current_selectors=current_selectors,
            failing_selectors=failing_selectors,
            html_sample=html_sample[:5000],
            page_type=page_type,
        )


class PluginDesignerModule(dspy.Module):
    """ChainOfThought wrapper for plugin design."""

    def __init__(self) -> None:
        super().__init__()
        self.design = dspy.ChainOfThought(PluginDesigner)

    def forward(
        self,
        domain: str,
        crawled_summaries: str,
        discovered_page_types: str = "",
        existing_plugins: str = "",
    ) -> dspy.Prediction:
        """Design a plugin from crawled knowledge."""
        return self.design(
            domain=domain,
            crawled_summaries=crawled_summaries,
            discovered_page_types=discovered_page_types,
            existing_plugins=existing_plugins,
        )


class CodegenRouterModule(dspy.Module):
    """ChainOfThought wrapper for code generation routing."""

    def __init__(self) -> None:
        super().__init__()
        self.route = dspy.ChainOfThought(CodegenRouter)

    def forward(
        self,
        task_description: str,
        target_environment: str = "cli",
        preferred_languages: str = "auto",
        constraints: str = "",
    ) -> dspy.Prediction:
        """Route a codegen task to appropriate languages."""
        return self.route(
            task_description=task_description,
            target_environment=target_environment,
            preferred_languages=preferred_languages,
            constraints=constraints,
        )
